src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContextSerializer.cs

```diff
@@ -25,25 +38,175 @@ private DesigntimeLicenseContextSerializer()
         /// </summary>
         public static void Serialize(Stream o, string cryptoKey, DesigntimeLicenseContext context)
         {
-            IFormatter formatter = new BinaryFormatter();
-#pragma warning disable SYSLIB0011 // Issue https://github.com/dotnet/runtime/issues/39293 tracks finding an alternative to BinaryFormatter
-            formatter.Serialize(o, new object[] { cryptoKey, context._savedLicenseKeys });
+            if (EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization)
+            {
+                SerializeWithBinaryFormatter(o, cryptoKey, context);
+            }
+            else
+            {
+                using (BinaryWriter writer = new BinaryWriter(o, encoding: Text.Encoding.UTF8, leaveOpen: true))
+                {
+                    writer.Write((sbyte)-1); // flag to identify BinaryWriter
+                    writer.Write(cryptoKey);
+                    writer.Write(context._savedLicenseKeys.Count);
+                    foreach (DictionaryEntry keyAndValue in context._savedLicenseKeys)
+                    {
+                        writer.Write(keyAndValue.Key.ToString());
+                        writer.Write(keyAndValue.Value.ToString());
+                    }
+                }
+            }
+        }
+
+        private static void SerializeWithBinaryFormatter(Stream o, string cryptoKey, DesigntimeLicenseContext context)
+        {
+            try
+            {
+                IFormatter formatter = new BinaryFormatter();
+#pragma warning disable SYSLIB0011
+                formatter.Serialize(o, new object[] { cryptoKey, context._savedLicenseKeys });
 #pragma warning restore SYSLIB0011
+            }
+            catch (NotSupportedException exception)
+            {
+                if (!BinaryFormatterSerializationEnabled)
+                {
+                    throw new NotSupportedException(exception.Message + " Turn on the EnableUnsafeBinaryFormatterSerialization flag to continue using BinaryFormatter");
+                }
+                else
+                {
+                    throw;
+                }
+            }
         }
 
-        internal static void Deserialize(Stream o, string cryptoKey, RuntimeLicenseContext context)
+        internal class StreamWrapper : Stream
         {
-#pragma warning disable SYSLIB0011 // Issue https://github.com/dotnet/runtime/issues/39293 tracks finding an alternative to BinaryFormatter
-            IFormatter formatter = new BinaryFormatter();
+            private Stream _stream;
+            internal bool readFirstByte;
```

Comment:
This doesn't appear to need to be `internal`. Can be private, and use the normal naming `_readFirstByte`.